src/common/transform.h

```diff
@@ -0,0 +1,211 @@
+ *     will merge functions with same signature.
+ */
+template <bool CompiledWithCuda = WITH_CUDA()>
+class Transform {
+ private:
+  template <typename Functor>
+  struct Evaluator {
+   public:
+    Evaluator(Functor func, Range range, GPUSet devices, bool reshard) :
+        func_(func), range_{range}, reshard_{reshard},
+        distribution_{GPUDistribution::Block(devices)} {}
+    Evaluator(Functor func, Range range, GPUDistribution dist, bool reshard):
+        func_(func), range_{range}, reshard_{reshard}, distribution_{dist} {}
+
+    /*!
+     * \brief Evaluate the functor with input pointers to HostDeviceVector.
+     *
+     * \tparam HDV...  HostDeviceVectors type.
+     * \param  vectors Pointers to HostDeviceVector.
+     */
+    template <typename... HDV>
+    void Eval(HDV... vectors) const {
+      bool on_device = !distribution_.IsEmpty();
+
+      if (on_device) {
+        LaunchCUDA(func_, vectors...);
+      } else {
+        LaunchCPU(func_, vectors...);
+      }
+    }
+
+   private:
+    // CUDA UnpackHDV
+    template <typename T>
+    Span<T> UnpackHDV(HostDeviceVector<T>* _vec, int _device) const {
+      return _vec->DeviceSpan(_device);
+    }
+    template <typename T>
+    Span<T const> UnpackHDV(const HostDeviceVector<T>* _vec, int _device) const {
+      return _vec->ConstDeviceSpan(_device);
+    }
+    // CPU UnpackHDV
+    template <typename T>
+    Span<T> UnpackHDV(HostDeviceVector<T>* _vec) const {
+      return Span<T> {_vec->HostPointer(),
+            static_cast<typename Span<T>::index_type>(_vec->Size())};
+    }
+    template <typename T>
+    Span<T const> UnpackHDV(const HostDeviceVector<T>* _vec) const {
+      return Span<T const> {_vec->ConstHostPointer(),
+            static_cast<typename Span<T>::index_type>(_vec->Size())};
+    }
+    // Recursive unpack for Reshard.
+    template <typename T>
+    void UnpackReshard(GPUDistribution dist, const HostDeviceVector<T>* vector) const {
+      vector->Reshard(dist);
+    }
+    template <typename Head, typename... Rest>
+    void UnpackReshard(GPUDistribution dist,
+                       const HostDeviceVector<Head>* _vector,
+                       const HostDeviceVector<Rest>*... _vectors) const {
+      _vector->Reshard(dist);
+      UnpackReshard(dist, _vectors...);
+    }
+
+#if defined(__CUDACC__)
+    template <typename std::enable_if<CompiledWithCuda>::type* = nullptr,
+              typename... HDV>
+    void LaunchCUDA(Functor _func, HDV*... _vectors) const {
+      if (reshard_)
+        UnpackReshard(distribution_, _vectors...);
+
+      GPUSet devices = distribution_.Devices();
+      size_t range_size = *range_.end() - *range_.begin();
+#pragma omp parallel for schedule(static, 1) if (devices.Size() > 1)
+      for (omp_ulong i = 0; i < devices.Size(); ++i) {
+        int d = devices.Index(i);
+        // Ignore other attributes of GPUDistribution for spliting index.
+        size_t shard_size =
```

Comment:
Could you use the shard size derived from `distribution_`? The `HostDeviceVector` objects are not necessarily block-distributed.

Reply:
Sorry about that, I made the changes in a local branch but haven't done the push yet. :)

Reply:
My mistake,  I tried to do that but gave up, since vectors are not necessarily come with the same distribution (different granularity for example).  The `shard_size` is defined for indexing thread, which should be fit the block distribution.